### ISSUE TITLE
Goals-model implementation landscape: Dixon-Coles, bivariate Poisson, XGBoost

### ISSUE THREAD

**Issue by mattyorkilous** [mattyorkilous/world-cup-predictor#3]
## Question

The model shape is settled — one unified goals model, results falling out of the scoreline. The implementation route is not.

- Is there a maintained Python library for Dixon-Coles / bivariate Poisson in 2026, or is hand-rolling the MLE the realistic path? Evaluate what exists against: actively maintained, works on 50k matches, supports sample weights (needed for era decay × importance).
- What is the standard low-score correction (the Dixon-Coles tau adjustment for 0-0/1-0/0-1/1-1) and does it still earn its place in current practice?
- How is time decay conventionally applied in this literature — and does it corroborate or contradict an ~8-year half-life?
- For **XGBoost with `count:poisson`**: how do practitioners handle two correlated outputs (home goals, away goals)? Two independent models, or something that preserves correlation?
- What does the current football-forecasting field actually use? Check what published approaches (Opta, FiveThirtyEight legacy methodology, academic papers) settled on.

Part of #1

**Comment by mattyorkilous** [comment]
## Findings

Full write-up with citations: [`docs/research/goals-model-landscape.md`](https://github.com/mattyorkilous/world-cup-predictor/blob/research/goals-model-landscape/docs/research/goals-model-landscape.md) on branch `research/goals-model-landscape`.

The single finding that runs through all five questions: **the model class barely matters; the features and the weighting do.** Three independent backtests put logistic regression, Poisson, XGBoost and ensembles within ~0.01 RPS of each other, while one feature change moved RPS by 0.08.

---

### 1. Library or hand-roll?

**Library — [penaltyblog](https://github.com/martineastwood/penaltyblog), and it is the only one.** A GitHub search for Python Dixon-Coles repos pushed since mid-2025 returns only one-off personal predictors (top: 32 stars), no libraries. `goalmodel` is R.

Against the three gates:

| Gate | Verdict |
|---|---|
| Maintained | **Pass.** MIT, 227 stars, 0 open issues, v1.12.2 released 2026-09-13, releases every 4–8 weeks through 2025–26 |
| 50k matches | **Pass.** I measured it: **3.2 s** for 50k matches / 250 teams / 502 params. Cython loss + analytic gradient into `scipy.optimize.minimize`. (Bivariate Poisson: 36.4 s.) |
| Per-sample weights | **Pass, first-class.** `weights` is a constructor arg, applied per match inside the likelihood: `total_llk += ((llk_home + llk_away) + adjustment) * weights[i]` |

Bonus not in the docs, found in the source: **`neutral_venue` is a per-match 0/1 flag** that drops home advantage for that match — `lambda_home = exp(hfa * (1 - neutral_venue[i]) + ...)`. For an international dataset that is exactly right, and it is the part that would be most annoying to retrofit onto a hand-roll.

**The one real limitation, and it matters here:** penaltyblog's goal models take team identities and weights, and **no per-match covariates**. There is no slot for Elo. #1 settled that Elo stays as a feature alongside attack/defence strengths — that decision cannot be expressed through this API.

So the route is: **library first; hand-roll when Elo-as-covariate forces it; keep the library installed as the characterization-test oracle** (fit both on the same data, assert the same log-likelihood). The hand-roll is ~40 lines, and penaltyblog's `gradients.pyx` is a line-by-line reference for the only fiddly part. That is the strangler pattern #1 already picked.

### 2. The tau correction — take it, expect nothing from it

One parameter shifting mass between four cells (0-0, 0-1, 1-0, 1-1), leaving everything else as independent Poissons. Exact form in the doc.

Does it still earn its place? Evidence, in increasing order of discomfort:

- **It wins, by 0.0001 RPS.** The package author's own rolling Eredivisie backtest: Dixon-Coles 0.1914, Weibull 0.1914, Poisson 0.1915, ZIP 0.1915, NegBin 0.1916, **bivariate Poisson 0.1916 (last)**. For scale, 538 reports that *adding xG to SPI* was worth 0.0018 RPS. Tau is a twentieth of that.
- **The pattern it corrects has drifted.** [arXiv:2103.07272](https://arxiv.org/abs/2103.07272) replicated Dixon & Coles' own diagnostic on 9,130 top-five-league matches 2014–19: the 0-0 cell sits at 99.47% of independence and is **not significant**. The big deviations are in the high-score corners (0-4 at 139.6, 4-4 at 52.6) — precisely where tau does nothing.
- **A purpose-built replacement beat it at p = 0.14.** Same paper. Their words: "it cannot be excluded that the two models are identical in terms of performance."

**Verdict: keep it because it is free in penaltyblog. Don't build anything around it, and don't count on it.** Caveat: tau is calibrated on club football; international has a fatter blowout tail, and our headline test is World Cups.

### 3. Time decay — the 8-year guess is contradicted

Convention is `w = (1/2)^(days_ago / HalfPeriod)`. Two independent studies on **international** football converge on **3 years**:

- **[Ley, Van de Wiele & Van Eetvelde (2019)](https://arxiv.org/abs/1705.09575)** — 4,268 competitive internationals 2008–17, weighted MLE with weights = time decay × match importance (i.e. exactly our scheme). Half period searched 0.5→6 years. Optimum **3 years** for both bivariate (RPS 0.1651) and independent Poisson (0.1653). For the Premier League in the same paper: 390 / 360 days.
- **[Groll et al. (2018)](https://arxiv.org/abs/1806.03208)** — all internationals 2002–17. Ranking: 1095 d > 730 d > 1460 d > … > 1825 d > 365 d. RPS is already degrading by 5 years.

Dixon-Coles' own finding, for the record: ξ = 0.0065 per *half-week* on 1990s English club data → ξ/day = 0.00186 → **half-life ≈ 373 days, about one year**. That is what penaltyblog's `xi=0.0018` default encodes. The club/international gap is real and is roughly a factor of three.

**Nothing supports 8 years.** Likely origin of the number, worth naming so it doesn't come back: *both* papers estimate over an **eight-year training window** — "parameters are estimated by maximum likelihood on a period of eight years". Eight years is the window, three is the half-life.

Two actions:
- **Default 3 years, tune 1–6** in the walk-forward backtest. Treat >6 as out of family.
- **Adopt Ley et al.'s importance weights** instead of inventing a scale — they are FIFA's and they're published: friendly **1**, qualifier **2.5**, confederation tournament / Confeds Cup **3**, World Cup **4**.

One consequence for the perf/caching work in #1: at a 3-year half-life an 1872 match carries weight 2^-51 ≈ 4e-16. Training "since 1872" is nominal — the effective training set is ~15 years deep. Worth phrasing honestly in the README too.

### 4. XGBoost `count:poisson` and two correlated outputs

**XGBoost has no joint bivariate count objective and won't get one soon.** `count:poisson` is univariate (`max_delta_step` defaults to 0.7 for it). Multi-output "builds one model for each target", is experimental as of 3.0 with "limited support from objectives and metrics", and `multi_output_tree` shares tree *structure*, not a joint likelihood.

**Does the discarded correlation matter? For what we score — essentially no, and there's a proof.** Ley et al. §2.3.2: under the bivariate Poisson, `G_i = X_i + X_C`, `G_j = X_j + X_C`, so

```
P(G_i − G_j = x) = P(X_i + X_C − (X_j + X_C) = x) = P(X_i − X_j = x)
```

**The shared component cancels out of the goal difference.** The bivariate Poisson's correlation parameter cannot change win/draw/loss probabilities *at all* — it only moves mass between scorelines of the same margin. Given #1 settled that results fall out of the scoreline, this is the load-bearing fact in the whole ticket.

Corroborating: Groll et al. note the bivariate Poisson "can only account for (positive) dependencies" while football's low-score dependence is negative, and report a boosting model selection where "the dependency parameter of the bivariate Poisson distribution was never updated by the algorithm". penaltyblog's backtest ranks it last. My benchmark: 11× the fit time.

Keep the asymmetry straight: **DC's tau *does* change 1X2** (it reshapes the draw cells relative to 1-0/0-1); **λ_C does not**. That's why tau survives in practice and the more principled-looking construction doesn't.

Practical shape for the XGBoost slate entry:
- **Long format, one model, not two** — one row per team-per-match with an `is_home` flag. Half the code, and the attack/defence signal is shared rather than split. `sample_weight` carries decay × importance unchanged.
- **Don't touch `multi_strategy`.**
- **If the grid needs a correlation fix, fix the grid, not the learner** — build the outer-product matrix from the two λs, then apply tau or a 538-style diagonal inflation downstream. A handful of lines, doesn't disturb training.

### 5. What published forecasters actually use

**FiveThirtyEight (legacy, archived):** SPI offence/defence ratings → two projected goal totals → **two independent Poissons** → score matrix → **~9% diagonal inflation** (league-specific). Verbatim: *"We've found that two independent Poisson distributions work well with the addition of diagonal inflation."* Not tau, not bivariate Poisson. Other contact points with #1: match importance is an explicit model input, and season forecasts are 20,000 Monte Carlo sims run "hot" — ratings move *within* a simulation, same as our mid-tournament Elo decision. Their own RPS ledger is the best yardstick available for "how big is a real improvement here": xG −0.0018, market values −0.0011, match importance −0.0006, final 0.1957.

**Opta:** publishes only "betting market odds + Opta Power Rankings → outcome probabilities → thousands of simulations". No goals model, no distributional assumption. Not reproducible, not a useful design reference — and market odds as an input is a different game from ours.

**Academia:**
- Ley et al. — bivariate Poisson with **one** strength parameter per team wins at both club and international level. Note the parsimony result: the single-parameter version beat separate attack/defence in both case studies.
- Groll et al. — a random forest alone roughly ties a Poisson ranking model (0.192 vs 0.190; bookmakers 0.188). Feed the Poisson model's **ability estimates in as a covariate** and the forest hits **0.187, beating the bookmakers**, with abilities "by far the most important predictor" ahead of FIFA rank, odds and GDP. Direct endorsement of #1's "Elo stays as a feature alongside attack/defence strengths".
- [Rezaei & Samadi (2026)](https://arxiv.org/abs/2606.24171) — backtested on **WC2018 + WC2022**, exactly our protocol. Elo-logistic 0.219, Poisson-on-Elo 0.212, XGBoost 0.215, ensemble 0.209 — their summary: *"Logistic regression, Poisson regression, ARIMA and NNAR variants, XGBoost, and the ensemble all perform similarly."* Then **0.127** from replacing current Elo with a dimension-reduced summary of its recent *history* — a feature change inside the same Poisson double-regression.

**Where the field landed:** a time-weighted double-Poisson goals model is the settled backbone, whether the λs come from attack/defence MLE, from Elo, or from a tree ensemble. Refinements to the joint distribution are academically interesting and empirically marginal. Gains come from features, weighting, and ensembling.

---

## Recommendation

1. **`penaltyblog.models.DixonColesGoalModel`** as the implementation. Clears all three gates plus gives us per-match neutral-venue handling for free. Zero MLE code to start.
2. **Hand-roll only when Elo-as-covariate forces it** — that's the known ceiling, ~40 lines, with penaltyblog retained as the test oracle.
3. **Skip the bivariate Poisson entirely.** Provably can't change 1X2, ranks last empirically, 11× slower.
4. **Keep tau.** Free, never hurts, expect ≈0.0001 RPS.
5. **Half-life 3 years, tuned over 1–6.** Drop 8. Adopt FIFA importance weights 1 / 2.5 / 3 / 4.
6. **XGBoost: long format, one `count:poisson` model, `is_home` flag, `sample_weight` = decay × importance.** No multi-output. Correlation fixes go on the grid, post hoc.

### What would change this
- penaltyblog going stale (one maintainer) — mitigated by the small hand-roll + oracle plan; re-check cadence before model work starts.
- A large, well-determined `rho` on the real international data — then tau is worth more than 0.0001 here.
- Needing exact-score or over/under outputs (not in scope) — "correlation doesn't matter" holds for 1X2 and margin, **not** for exact scorelines.

**Comment by mattyorkilous** [comment]
Resolved — findings posted above, committed to `docs/research/goals-model-landscape.md` on branch `research/goals-model-landscape`.
